File: Solution/0098.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool ValidBST(TreeNode* root, long left_min, long right_max){
        if(root == nullptr) return true;
        if(root -> val >= left_min || root -> val <= right_max) return false;

        return ValidBST(root -> left, root -> val, right_max) && 
               ValidBST(root -> right, left_min, root -> val);        

    }
    bool isValidBST(TreeNode* root) {
        return ValidBST(root, LONG_MAX, LONG_MIN);
    }
};

File: Solution/0098.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def checkBST(self, root, low_min, high_max):
        if root == None:
            return True

        if root.val <= low_min or root.val >= high_max:
            return False
        
        return self.checkBST(root.left, low_min, root.val) and self.checkBST(root.right, root.val, high_max)
        
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.checkBST(root, -float("inf"), float("inf"))